### For Web Scraping we need "Internet-Connetion "

In [1]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup

In [2]:
requests.get('https://www.ambitionbox.com/list-of-companies?page=1')

<Response [403]>

### so we found "Respose [403] " it means that the site is not giving us permition so

In [3]:
requests.get('https://www.ambitionbox.com/list-of-companies?page=1').text

'<HTML><HEAD>\n<TITLE>Access Denied</TITLE>\n</HEAD><BODY>\n<H1>Access Denied</H1>\n \nYou don\'t have permission to access "http&#58;&#47;&#47;www&#46;ambitionbox&#46;com&#47;list&#45;of&#45;companies&#63;" on this server.<P>\nReference&#32;&#35;18&#46;1ef63717&#46;1669450711&#46;abf2979\n</BODY>\n</HTML>\n'

### Here we see that the site thing we are "Bot" so We make site beleave that we are human

In [4]:
headers={'User-Agent':'Mozilla/5.0 (Windows NT 6.3; Win 64 ; x64) Apple WeKit /537.36(KHTML , like Gecko) Chrome/80.0.3987.162 Safari/537.36'}

In [5]:
webpage=requests.get('https://www.ambitionbox.com/list-of-companies?page=1',headers = headers).text

### For Understanding data we need to use "BeautifulSoup - laibrary " and go to page to able to understand data and Watch "100Days-of-ML" in YouTube , Video Name "Feching data using Web scaping"

In [6]:
soup = BeautifulSoup(webpage,'lxml') # Making this data little bit readable

### for extracting values from this data we have to find valuable data so we need to go site and click ringh on the mouse and select inspect and then find botten  and click it help to find data  

In [7]:
soup.find_all('h1')

[<h1 class="bold-title-l title">List of companies in India</h1>]

##### so need text from this list

In [8]:
soup.find_all('h1')[0].text #we need indexing

'List of companies in India'

### We need Company name

In [9]:
len(soup.find_all('h2')) # we use len to find the leght of company names

30

In [10]:
soup.find_all('h2')

[<h2 class="company-name bold-title-l" title="TCS">
 									TCS
 								</h2>,
 <h2 class="company-name bold-title-l" title="Accenture">
 									Accenture
 								</h2>,
 <h2 class="company-name bold-title-l" title="Cognizant">
 									Cognizant
 								</h2>,
 <h2 class="company-name bold-title-l" title="ICICI Bank">
 									ICICI Bank
 								</h2>,
 <h2 class="company-name bold-title-l" title="HDFC Bank">
 									HDFC Bank
 								</h2>,
 <h2 class="company-name bold-title-l" title="Wipro">
 									Wipro
 								</h2>,
 <h2 class="company-name bold-title-l" title="Infosys">
 									Infosys
 								</h2>,
 <h2 class="company-name bold-title-l" title="Capgemini">
 									Capgemini
 								</h2>,
 <h2 class="company-name bold-title-l" title="Tech Mahindra">
 									Tech Mahindra
 								</h2>,
 <h2 class="company-name bold-title-l" title="Genpact">
 									Genpact
 								</h2>,
 <h2 class="company-name bold-title-l" title="HCLTech">
 									HCLTech
 			

#### so need text from this list

#### there is so many values so we need Loop

In [11]:
for i in soup.find_all('h2') :
    print(i.text)


									TCS
								

									Accenture
								

									Cognizant
								

									ICICI Bank
								

									HDFC Bank
								

									Wipro
								

									Infosys
								

									Capgemini
								

									Tech Mahindra
								

									Genpact
								

									HCLTech
								

									Amazon
								

									Axis Bank
								

									IBM
								

									Concentrix Corpo...
								

									Reliance jio
								

									Larsen & Toubro ...
								

									Vodafone Idea
								

									HDB Financial Se...
								

									Teleperformance
								

									Kotak Mahindra B...
								

									Reliance Industr...
								

									Bharti Airtel
								

									Reliance Retail
								

									Deloitte
								

									Tata Motors
								

									BYJU'S
								

									WNS
								

									Mahindra & Mahin...
								

									Flipkart
								


### we Extract values but there is problem in python we do not able to se but there is HTML "Space And / " are there and we need to put this values in manner ways So need to Use "Strip()" funtion

In [12]:
for i in soup.find_all('h2'):
    print(i.text.strip())

TCS
Accenture
Cognizant
ICICI Bank
HDFC Bank
Wipro
Infosys
Capgemini
Tech Mahindra
Genpact
HCLTech
Amazon
Axis Bank
IBM
Concentrix Corpo...
Reliance jio
Larsen & Toubro ...
Vodafone Idea
HDB Financial Se...
Teleperformance
Kotak Mahindra B...
Reliance Industr...
Bharti Airtel
Reliance Retail
Deloitte
Tata Motors
BYJU'S
WNS
Mahindra & Mahin...
Flipkart


### We need Reviews of Company

In [13]:
len(soup.find_all('a'))

428

### so we find it that the lenth of this Review is more than 30 so need to go the web page and find a and its class name

In [14]:
len(soup.find_all('a',class_='review-count sbold-Labels'))

30

In [15]:
for i in soup.find_all('a',class_ = 'review-count sbold-Labels'):
    print(i.text.strip())

(50.5k Reviews)
(32.8k Reviews)
(30k Reviews)
(35.2k Reviews)
(40.3k Reviews)
(29.1k Reviews)
(28.6k Reviews)
(24.3k Reviews)
(22.1k Reviews)
(21.3k Reviews)
(18.8k Reviews)
(18.6k Reviews)
(19.3k Reviews)
(15.8k Reviews)
(15.3k Reviews)
(15.2k Reviews)
(25.1k Reviews)
(14k Reviews)
(13.5k Reviews)
(14.6k Reviews)
(13.9k Reviews)
(44.5k Reviews)
(13k Reviews)
(16k Reviews)
(10.4k Reviews)
(12.6k Reviews)
(12.4k Reviews)
(7.1k Reviews)
(40.1k Reviews)
(11.7k Reviews)


In [16]:
len(soup.find_all('p')) 

229

### Here Our apprach is wronge, Why! , we need to find container  so we need to go to web page and click botten in find a box which contain all the values like campany name , rating ,reviews ets but "We Need to understand Little-Bit of HTMl" 

In [17]:
company = soup.find_all('div',class_='company-content-wrapper')

In [18]:
len(company)

30

### here what we did is we extract values values from a perticuler container before we run the loop in whole HMTL columns  but now we run it into this only  Container 

In [19]:
name=[]

for i in company :
    name.append(i.find('h2').text.strip())

In [20]:
name

['TCS',
 'Accenture',
 'Cognizant',
 'ICICI Bank',
 'HDFC Bank',
 'Wipro',
 'Infosys',
 'Capgemini',
 'Tech Mahindra',
 'Genpact',
 'HCLTech',
 'Amazon',
 'Axis Bank',
 'IBM',
 'Concentrix Corpo...',
 'Reliance jio',
 'Larsen & Toubro ...',
 'Vodafone Idea',
 'HDB Financial Se...',
 'Teleperformance',
 'Kotak Mahindra B...',
 'Reliance Industr...',
 'Bharti Airtel',
 'Reliance Retail',
 'Deloitte',
 'Tata Motors',
 "BYJU'S",
 'WNS',
 'Mahindra & Mahin...',
 'Flipkart']

### so we find find good way to extract valueable values

In [21]:
a=company[0].find_all('p',class_='infoEntity')[0].text.strip()
b=company[1].find_all('p',class_='infoEntity')[1].text.strip()
c=company[2].find_all('p',class_='infoEntity')[2].text.strip()
d=company[3].find_all('p',class_='infoEntity')[3].text.strip()

In [22]:
print(a)

Public


In [23]:
b

'Dublin + 142 more'

In [24]:
c

'28 years old'

In [25]:
d

'1 Lakh+ Employees (India)'

### so this has same class name that the mane reason we use this method and next we store them in different varialbles

In [26]:
name=[]
rating=[]
reviews=[]
ctype=[]
hq=[]
how_old=[]
no_of_employee=[]
for i in company :
    name.append(i.find('h2').text.strip()) #we use find here because this perticular container has only 1 "h2" value
    rating.append(i.find('p',class_= 'rating').text.strip())
    reviews.append(i.find('a',class_= 'review-count sbold-Labels').text.strip())
    ctype.append(i.find_all('p',class_='infoEntity')[0].text.strip())
    hq.append(i.find_all('p',class_ = 'infoEntity')[1].text.strip())
    how_old.append(i.find_all('p',class_ = 'infoEntity')[2].text.strip())
    no_of_employee.append(i.find_all('p',class_ = 'infoEntity')[3].text.strip())

### Now we need to store these values in DataFrame


In [27]:
df = pd.DataFrame({"name":name,"rating":rating,"reviews":reviews,"Company_type":ctype,"Head_Quator":hq,"Company_age":how_old,"No_of_Employees":no_of_employee})

In [28]:
df

name rating          reviews Company_type  \
0                   TCS    3.9  (50.5k Reviews)       Public   
1             Accenture    4.1  (32.8k Reviews)       Public   
2             Cognizant    3.9    (30k Reviews)      Private   
3            ICICI Bank    4.0  (35.2k Reviews)       Public   
4             HDFC Bank    4.0  (40.3k Reviews)       Public   
5                 Wipro    3.9  (29.1k Reviews)       Public   
6               Infosys    3.9  (28.6k Reviews)       Public   
7             Capgemini    3.9  (24.3k Reviews)       Public   
8         Tech Mahindra    3.7  (22.1k Reviews)       Public   
9               Genpact    4.0  (21.3k Reviews)       Public   
10              HCLTech    3.8  (18.8k Reviews)       Public   
11               Amazon    4.3  (18.6k Reviews)       Public   
12            Axis Bank    3.9  (19.3k Reviews)       Public   
13                  IBM    4.2  (15.8k Reviews)       Public   
14  Concentrix Corpo...    4.0  (15.3k Reviews)       Public   
15         Reliance jio    4.0  (15.2k Reviews)       Public   
16  Larsen & Toubro ...    4.1  (25.1k Reviews)       Public   
17        Vodafone Idea    4.2    (14k Reviews)       Public   
18  HDB Financial Se...    4.0  (13.5k Reviews)      Private   
19      Teleperformance    3.6  (14.6k Reviews)      Private   
20  Kotak Mahindra B...    3.9  (13.9k Reviews)       Public   
21  Reliance Industr...    4.1  (44.5k Reviews)       Public   
22        Bharti Airtel    4.1    (13k Reviews)       Public   
23      Reliance Retail    4.1    (16k Reviews)      Private   
24             Deloitte    4.1  (10.4k Reviews)      Private   
25          Tata Motors    4.1  (12.6k Reviews)       Public   
26               BYJU'S    3.5  (12.4k Reviews)      Private   
27                  WNS    3.7   (7.1k Reviews)      Private   
28  Mahindra & Mahin...    3.9  (40.1k Reviews)       Public   
29             Flipkart    4.2  (11.7k Reviews)       Public   

                                 Head_Quator    Company_age  \
0              Mumbai,Maharashtra + 278 more   54 years old   
1                          Dublin + 142 more   33 years old   
2            Teaneck. New Jersey. + 100 more   28 years old   
3             Mumbai,Maharashtra + 1046 more   28 years old   
4             Mumbai,Maharashtra + 1225 more   28 years old   
5   Bangalore/Bengaluru,Karnataka + 231 more   77 years old   
6   Bengaluru/Bangalore,Karnataka + 126 more   41 years old   
7                            Paris + 80 more   55 years old   
8                Pune,Maharashtra + 217 more   36 years old   
9                New York,New York + 67 more   25 years old   
10                          Noida + 142 more   31 years old   
11             Seattle,Washington + 387 more   28 years old   
12            Mumbai,Maharashtra + 1078 more   29 years old   
13                Armonk,New York + 122 more  111 years old   
14              Fremont,California + 71 more   39 years old   
15        Navi Mumbai,Maharashtra + 969 more   15 years old   
16             Mumbai,Maharashtra + 525 more   84 years old   
17             Gandhinagar,Gujrat + 564 more    4 years old   
18               Ahmedabad,Gujrat + 780 more   15 years old   
19                          Paris + 124 more   44 years old   
20             Mumbai,Maharashtra + 472 more   19 years old   
21        Navi Mumbai,Maharashtra + 488 more   49 years old   
22       Gurgaon/Gurugram,Haryana + 512 more   27 years old   
23        Navi Mumbai,Maharashtra + 659 more   16 years old   
24              New York,New York + 134 more  177 years old   
25               Pune,Maharashtra + 394 more   77 years old   
26            Bangalore,Karnataka + 215 more   11 years old   
27              Mumbai,Maharashtra + 23 more   26 years old   
28             Mumbai,Maharashtra + 410 more   77 years old   
29            Bangalore,Karnataka + 441 more   15 years old   

                 No_of_Employees  
0      1 Lakh+ Employees (India)  
1      1 L

# creating dataframe for all the pages

In [29]:
final_data = pd.DataFrame()

In [30]:
for j in range(1,11):
    webpage=requests.get('https://www.ambitionbox.com/list-of-companies?page={}',format(j)).text
    soup = BeautifulSoup(webpage,'lxml') # Making this data little bit readable
    company=soup.find_all('div',class_='company-content-wrapper')
    
    
    name=[]
    rating=[]
    reviews=[]
    ctype=[]
    hq=[]
    how_old=[]
    no_of_employee=[]
    for i in company:
        try:
           name.append(i.find('h2').text.strip())
        except:
           name.append(np.nan)

        try:
           rating.append(i.find('p',class_='rating').text.strip())
        except:
           rating.append(np.nan)

        try:

          reviews.append(i.find('a' , class_='review-count').text.strip())
        except:
          reviews.append(np.nan)

        try:

          ctype.append(i.find_all('p',class_='infoEntity')[0].text.strip())
        except:
          ctype.append(np.nan)
        try:

          hq.append(i.find_all('p',class_='infoEntity')[1].text.strip())
        except:
          hq.append(np.nan)

        try:

          how_old.append(i.find_all('p',class_='infoEntity')[2].text.strip())
        except:
          how_old.append(np.nan)
        try:
          no_of_employee.append(i.find_all('p',class_='infoEntity')[3].text.strip())
        except:
          no_of_employee.append(np.nan)
        
        
        
        

    df = pd.DataFrame({"name":name,"rating":rating,"reviews":reviews,"Company_type":ctype,"Head_Quator":hq,"Company_age":how_old,"No_of_Employees":no_of_employee})
    final_data = final_data.append(df,ignore_index=True)
    

C:\Users\hites\AppData\Local\Temp\ipykernel_13544\3388097988.py:57: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  final_data = final_data.append(df,ignore_index=True)


In [31]:
final_data.shape

(0, 7)

In [32]:
df

Empty DataFrame
Columns: [name, rating, reviews, Company_type, Head_Quator, Company_age, No_of_Employees]
Index: []